### PR TITLE
Put all resources in specified provider namespace

### LINKED ISSUE #415: Provider namespace field doesn't put new resources in the namespace
I'm not sure if this is intentional, but here's the repro:

```
const k8s = require("@pulumi/kubernetes");

const prov = new k8s.Provider("prov", {
    namespace: "mynamespace",
});

const nginx = new k8s.core.v1.Pod("nginx", {
    spec: {
        containers: [{
            name: "nginx",
            image: "nginx",
        }]
    }
}, { provider: prov });

exports.namespace = nginx.metadata.apply(m => m.namespace);
```

This program creates a `Pod` in the `default` namespace, where I'd expect it to be put in `mynamespace`. If this is intentional, let's document it better, since it kinda seems like the `namespace` argument on the provider applies a default namespace of `mynamespace` if no other namespace is provided.

### PR DESCRIPTION
**Pull request by lblackstone**
This is an updated version of #506 that should correctly handle CRs.

Fixes #415

### CODE REVIEW

**Comment by lblackstone** [comment]
Error is:
```
kubernetes:config.istio.io:rule istio-system/promhttp  error: kubernetes:config.istio.io/v1alpha2:rule resource 'istio-system/promhttp' has a problem:  Found API Group, but it did not contain a schema for "config.istio.io/v1alpha2, Kind=rule"
```

Not sure yet if that's legit, or if https://github.com/pulumi/pulumi-kubernetes/pull/538/files#diff-6155820fac2dd2fda7353505b1f69319R307 is invalid.

File: pkg/clients/clients.go [hunk 1]
```diff
@@ -132,10 +132,15 @@ func (dcs *DynamicClientSet) gvkForKind(kind kinds.Kind) (*schema.GroupVersionKi
 	return nil, fmt.Errorf("failed to find gvk for Kind: %q", kind)
 }
 
-func (dcs *DynamicClientSet) namespaced(gvk schema.GroupVersionKind) (bool, error) {
-	resourceList, err := dcs.DiscoveryClientCached.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
+func (dcs *DynamicClientSet) NamespacedKind(gvk schema.GroupVersionKind) (bool, error) {
```
Comment (metral) [inline pkg/clients/clients.go]: nit: s/NamespacedKind/IsNamespacedKind

**Comment by lblackstone** [comment]
@metral I think I addressed all your feedback. Added a note to the code clarifying the override behavior.

CRs are a special case because it isn't always possible to tell if they are Namespaced/Global scoped during preview. As noted in the code, this could result in a namespace being set erroneously on a Cluster-scoped CR for previews, but this would be corrected during the Create step, when the required information is available. This correction will be reflected in further updates/diffs as well, so even if the erroneous namespace on the resource changes, it will not require a replacement.

**Comment by lblackstone** [comment]
@hausdorff Have you had a chance to look this over? The main change from the previously-merged #506 is the improved logic related to unregistered CRs. See the explanation in my comment to Mike for details.

**Pull request merged**